### PR TITLE
OPER-3354 Add ability to update existing ASGs w/ current config

### PR DESCRIPTION
**Pull request by alanbrent**
More description to come when things are closer to done.

What I envision, once we've done a one-time sweep to ensure all currently-in-production ASG settings are reflected accurately in the `.tass` YAMLs, is something like this:

Update all `tapjoyserver` ASGs:

```bash
$ cd $DOT_TASS_REPO
$ for i in $(find config/clusters -name 'tapjoyserver-*'); do
    tass update -f $i # if this were fully automated it would be called w/ --no-prompt
  done
```

### CODE REVIEW

**Comment by alanbrent** [comment]
As it stands at this moment, this code will update ASGs via `tass update -f $cfgfile`

File: lib/tapjoy/autoscaling_bootstrap.rb [hunk 1]
```diff
@@ -210,8 +210,8 @@ def legacy_configure_environment(opts)
 
         new_config[:autoscale] = false unless new_config[:scaling_type].eql?('dynamic')
         new_config[:tags] << {Name: new_config[:name]}
-        Tapjoy::AutoscalingBootstrap.scaler_name = "#{new_config[:name]}-group"
-        Tapjoy::AutoscalingBootstrap.config_name = "#{new_config[:name]}-config"
+        Tapjoy::AutoscalingBootstrap.scaler_name ||= "#{new_config[:name]}-group"
```
Comment (alanbrent) [inline lib/tapjoy/autoscaling_bootstrap.rb]: These variables were getting overwritten.

File: lib/tapjoy/autoscaling_bootstrap/launch_configuration.rb [hunk 2]
```diff
@@ -27,8 +27,8 @@ def current
       def update(config_name:, scaler_name: 'NaS', user_data: user_data,
         updated_config:, aws_env:)
 
-        Tapjoy::AutoscalingBootstrap.config_name = config_name
-        Tapjoy::AutoscalingBootstrap.scaler_name = scaler_name
+        Tapjoy::AutoscalingBootstrap.config_name ||= config_name
```
Comment (alanbrent) [inline lib/tapjoy/autoscaling_bootstrap/launch_configuration.rb]: As stated above, these variables were being overwritten.

**Comment by alanbrent** [comment]
Current output of the command is pretty much exactly the same as `tass create`:

```bash
$ bundle exec ./bin/tass update -f ../dot_tass/config/clusters/tapjoyserver-jobs_default_vpc.yaml 
NOTE: --env has been deprecated.  Please specify environment in yaml configs.
  Preparing to configure the following autoscaling group:
  Launch Config:        tapjoyserver-jobs_default_vpc-config
  Auto Scaler:          tapjoyserver-jobs_default_vpc-group
  Key Pair:             ops-keypair
  Zones:                us-east-1b,us-east-1d,us-east-1e
  Groups:               sg-6fefa00b
  Instance Type:        r3.xlarge
  Image ID:             ami-279ffc31
  IAM Role:             ec2_general_purpose
  VPC Subnets:          subnet-af6ad384,subnet-c8a6d5bf,subnet-6971e730
  Termination Policies: OldestInstance


NOTE! Continuing may have adverse effects if you end up deleting an IN-USE PRODUCTION scaling group. Don't be dumb.
Is this information correct? [y/n]
y
Updating launch config for scaling group: tapjoyserver-jobs_default_vpc-group
Current launch config is: tapjoyserver-jobs_default_vpc-config_20170503-161644

Launch config tapjoyserver-jobs_default_vpc-config_20170721-234729 does not exist, continuing...
Creating launch config: tapjoyserver-jobs_default_vpc-config_20170721-234729
Creating autoscale alerts and policies
Clearing out original CloudWatch alarms
Deleting alarm: tapjoyserver-jobs_default_vpc - Scale Up - High CPU Utilization
Deleting alarm: tapjoyserver-jobs_default_vpc - Scale Down - Low CPU Utilization

Clearing out original scaling policies
Deleting policy: tapjoyserver-jobs_default_vpc-up-policy
Deleting policy: tapjoyserver-jobs_default_vpc-down-policy

Configuring autoscaling...
Creating scale policy: tapjoyserver-jobs_default_vpc-up-policy
Creating alarms for scale policy: arn:aws:autoscaling:us-east-1:266171351246:scalingPolicy:6c9b6307-4acb-4d65-a2f1-8db0378b2562:autoScalingGroupName/tapjoyserver-jobs_default_vpc-group:policyName/tapjoyserver-jobs_default_vpc-up-policy
Creating: tapjoyserver-jobs_default_vpc - Scale Up - High CPU Utilization
Creating scale policy: tapjoyserver-jobs_default_vpc-down-policy
Creating alarms for scale policy: arn:aws:autoscaling:us-east-1:266171351246:scalingPolicy:5a1d56e4-bfe1-4002-94a3-d1362ebac53c:autoScalingGroupName/tapjoyserver-jobs_default_vpc-group:policyName/tapjoyserver-jobs_default_vpc-down-policy
Creating: tapjoyserver-jobs_default_vpc - Scale Down - Low CPU Utilization


Created -- 
Autoscaling Group: tapjoyserver-jobs_default_vpc-group
Launch Config:     tapjoyserver-jobs_default_vpc-config_20170721-234729
Instance Count:    0
```

**Comment by alanbrent** [comment]
@Tapjoy/eng-group-ops Although I've been able to get this to a point where one can apply updates to existing ASGs, I need some help here. The overall code flow still appears fairly opaque/circuitous to me, and I’d love to know what the correct modification points are at the following levels: 

• I'd love to stop re-using[`#create`](https://github.com/Tapjoy/tass/pull/43/files#diff-0aa418b3b15c71736c97badc228d9706) as the CLI target when what I want is an update-specific code flow that can be trusted/tested
    - one reason a separate flow is desirable would be to avoid tearing down all the related AWS constructs  `update` is issued if they haven't changed. which is the current behavior.
• branching from the immediately above: I am having difficulty untangling the precise relationship(s) of `Tapjoy::AutoscalingBootstrap::AutoscalingGroup.class` to the rest of the code flow supporting modifications . Any explanations would be **highly** welcome
• branching from the immediately above: I'm struggling to grok the layout of and naming in this project, and would love a TLDR

File: lib/tapjoy/autoscaling_bootstrap/configure_autoscaler.rb [hunk 3]
```diff
@@ -16,6 +20,11 @@ def initialize(misc_config:, aws_env:, user_data:, **unused_values)
 
       private
 
+      # Check if ASG exists so we know whether to create or update
+      def autoscaling_group_exists
+        !Tapjoy::AutoscalingBootstrap::AWS::Autoscaling::Group.describe.nil?
```
Comment (marcuswalser) [inline lib/tapjoy/autoscaling_bootstrap/configure_autoscaler.rb]: This reads like a double-negative and so is a little hard to parse. Can it be restated with `unless` or something?

Comment (deanxorr) [inline lib/tapjoy/autoscaling_bootstrap/configure_autoscaler.rb]: actually it looks like `activesupport` is included, so could it be replaced with `Tapjoy::AutoscalingBootstrap::AWS::Autoscaling::Group.describe.present?`

Comment (atayarani) [inline lib/tapjoy/autoscaling_bootstrap/configure_autoscaler.rb]: Dean's approach here is my preference as well

Comment (alanbrent) [inline lib/tapjoy/autoscaling_bootstrap/configure_autoscaler.rb]: I ripped this from https://github.com/Tapjoy/tass/blob/3fd8f74d9cac05a59bc1c9a56bac9e1b695d1c2c/lib/tapjoy/autoscaling_bootstrap/autoscaling/group.rb#L46-L48, so I'll update it there as well.

Comment (alanbrent) [inline lib/tapjoy/autoscaling_bootstrap/configure_autoscaler.rb]: Turns out activesupport is not included in this code flow, so I've reverted to using the previous code. Here are a couple of visual alternatives, which I don't find to be better in any compelling way:

```
Tapjoy::AutoscalingBootstrap::AWS::Autoscaling::Group.describe.nil? ? false : true
```

```
if Tapjoy::AutoscalingBootstrap::AWS::Autoscaling::Group.describe.nil?
  false
else
  true
end
# the unless form is no better
```

Comment (geoffreyclark) [inline lib/tapjoy/autoscaling_bootstrap/configure_autoscaler.rb]: My preference is for `Tapjoy::AutoscalingBootstrap::AWS::Autoscaling::Group.describe.nil? ? false : true` Stylistically you should consider changing the method name to `autoscaling_group_exists?` which indicates a boolean will be returned.

Comment (alanbrent) [inline lib/tapjoy/autoscaling_bootstrap/configure_autoscaler.rb]: These are also my preferences, and I'm happy to add them now that you stated them in a relatively unprompted manner :)
